assets/js/base/context/cart-checkout/checkout/processor/index.js

```diff
@@ -135,14 +141,16 @@ const CheckoutProcessor = () => {
 	] );
 
 	useEffect( () => {
-		const unsubscribeProcessing = onCheckoutProcessing(
-			checkValidation,
-			0
-		);
+		let unsubscribeProcessing;
+		if ( ! expressPaymentMethodActive ) {
+			unsubscribeProcessing = onCheckoutProcessing( checkValidation, 0 );
```

Comment:
In the case of the `onCheckoutProcessing` event emitter, I was thinking we'd want the checkout validator to run first. I could be wrong though, why were you thinking it should run last?  Also, does it matter if all observer return values are going to be merged (or is that _precisely the reason it should be last_, so checkout validation errors have priority)?

Reply:
I thought the `checkValidation` function as a wrap up that would make sure that we invalidate if something has errored, that's why it checks if `currentPaymentStatus.hasError` or `shippingErrorStatus.hasError` are true and returns an error in that case. Just to make sure we abort if any of those is in an error state.

But if we are confident shipping and payment methods will always hook up into `onCheckoutProcessing` and return an error there, it should be save to remove these lines, no?
https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/base/context/cart-checkout/checkout/processor/index.js#L113-L128

In that case, yes, probably running `checkValidation` with priority `0` as you said makes more sense so it runs at the beginning.

Reply:
With recent changes, payment processing happens on the `onPaymentProcessing` event which fires when the checkout status is `checkoutIsProcessingComplete`.  Which is set when checkout has successfully processed.

This change was needed so that _Payment Status_ could be controlled by checkout instead of payment methods and related to the fact that event observers cannot change context state.

However shipping error and payment errors might still get picked up by the logic you added in the linked lines because the checkout processing could be reverted to "IDLE" and have to be triggered again (this likely still will need adjusted).

I think we could leave things as is for now here priority wise and we can revisit if needed as we implement more things.